### PR TITLE
feat(cpu): Add FreeBSD support

### PR DESCRIPTION
**Pull request by Tectu**
## What type of PR is this?

* [ ] Refactor
* [x] Feature
* [ ] Bug Fix
* [ ] Optimization
* [ ] Documentation Update
* [ ] Other:

## Description
This adds support for the FreeBSD operating system to the CPU module.

## Related Issues & Documents
N/A

## Documentation (check all applicable)

* [ ] This PR requires changes to the Wiki documentation (describe the changes)
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [x] Does not require documentation changes

### CODE REVIEW

File: src/modules/cpu.cpp [hunk 1]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+      }
+
+      // Query 'kern.cp_time'. The returned values are (in order):
+      //  - user
+      //  - nice
+      //  - system
+      //  - interrupt
+      //  - idle
+      //
+      static constexpr std::size_t cpu_states_num = 5;
```
Comment (yuripv) [inline src/modules/cpu.cpp:157]: sys/resource.h: CPUSTATES

and constants below from the same header

File: src/modules/cpu.cpp [hunk 2]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+      if (sysctlbyname("kern.cp_time", cpu_stat, &sz, nullptr, 0) != 0) {
+        m_log.err("Failed to query sysctl 'kern.cp_time' (errno: %s)", strerror(errno));
+        return false;
+      }
+
+      // Parse
+      static std::size_t field_offset = sizeof(*cpu_stat) + ncpu;
+      for (std::size_t i = 0; i < ncpu; i++) {
+        m_cputimes.emplace_back(new cpu_time);
+        m_cputimes.back()->user = cpu_stat[0];
```
Comment (yuripv) [inline src/modules/cpu.cpp:169]: CP_USER

File: src/modules/cpu.cpp [hunk 3]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+        m_cputimes.back()->nice = cpu_stat[1];
```
Comment (yuripv) [inline src/modules/cpu.cpp:170]: CP_NICE

File: src/modules/cpu.cpp [hunk 4]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+        m_cputimes.back()->system = cpu_stat[2];
```
Comment (yuripv) [inline src/modules/cpu.cpp:171]: CP_SYS

File: src/modules/cpu.cpp [hunk 5]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+        m_cputimes.back()->steal = cpu_stat[3];   // Note: This is technically the reported "interrupt" time
```
Comment (yuripv) [inline src/modules/cpu.cpp:172]: CP_INTR

File: src/modules/cpu.cpp [hunk 6]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+        m_cputimes.back()->idle = cpu_stat[4];
```
Comment (yuripv) [inline src/modules/cpu.cpp:173]: CP_IDLE

File: src/modules/cpu.cpp [hunk 7]
```diff
@@ -129,6 +133,48 @@ namespace modules {
     m_cputimes.clear();
 
     try {
+#if __FreeBSD__
```
Comment (yuripv) [inline src/modules/cpu.cpp:136]: #ifdef would be more safe (and common)

File: src/modules/cpu.cpp [hunk 8]
```diff
@@ -129,6 +133,48 @@ namespace modules {
...
+      // Query 'kern.cp_time'. The returned values are (in order):
+      //  - user
+      //  - nice
+      //  - system
+      //  - interrupt
+      //  - idle
+      //
+      static constexpr std::size_t cpu_states_num = 5;
+      long cpu_stat[cpu_states_num];
+      sz = sizeof(*cpu_stat) * cpu_states_num;
```
Comment (yuripv) [inline src/modules/cpu.cpp:159]: Why not simply sizeof(cpu_stat)?

**Comment by Tectu** [comment]
@yuripv Thanks for your review - couldn't agree more!

**Comment by patrick96** [comment]
Thanks for the effort you put in here.

The current policy in polybar is to not add BSD-specific module code into our code base.
We have some BSD-specific code to make it compile without patches, but any actual BSD-specific implementation code should be maintained as patches in their respective packaging repo.

As discussed privately, this may change once we have a robust system for separating modules from their data-gathering code which would allow us to have different implementations for different systems that are accessed over the same API

**Pull request closed without merging**